src/bridge/crypto/rsa.cc

```diff
@@ -0,0 +1,250 @@
+// algorithm that was used to generate the signature. `rsa` is the signer's
+// public key.
+//
+// Returns 1 on success; otherwise, returns 0.
+//
+// The function signature comes from the prototype for `RSA_meth_set_verify`
+// from the OpenSSL API.
+int Verify(int type, const unsigned char *m, unsigned int m_len,
+           const unsigned char *sigbuf, unsigned int siglen, const RSA *rsa) {
+  // TODO(https://github.com/googleinterns/cloud-kms-oss-tools/issues/101):
+  // This method is currently purposely left unimplemented, but it may need to
+  // be implemented at some point in the future. The OpenSSL CLI seems to use
+  // RSA_public_decrypt over RSA_verify, so this never really gets called at the
+  // moment.
+  KMSENGINE_SIGNAL_ERROR(
+      Status(StatusCode::kUnimplemented, "Unsupported operation"));
+  return false;
+}
+
+// Signs the `from_length` bytes in `from` using the RSA key `rsa` and stores
+// the signature in `to`. The caller is responsible for ensuring that `to`
+// points to `RSA_size(rsa)` bytes of memory. (See the OpenSSL man page for
+// RSA_private_encrypt(3) for more information.)
+//
+// According to the OpenSSL specification, `from` is usually a message
+// digest, but it doesn't have to be.
+//
+// Returns the length of `to` on success; otherwise, returns -1.
+//
+// The function signature comes from the prototype for `RSA_meth_set_priv_enc`
+// from the OpenSSL API.
+int PrivateEncrypt(int from_length, const unsigned char *from,
+                   unsigned char *to, RSA *rsa, int padding) {
+  // TODO(https://github.com/googleinterns/cloud-kms-oss-tools/issues/100): It
+  // may or may not be possible to simply redirect `PrivateEncrypt` to `Sign`.
+  KMSENGINE_SIGNAL_ERROR(
+      Status(StatusCode::kUnimplemented, "Unsupported operation"));
+  return -1;
+}
+
+// Recovers the message digest from the `from_length` bytes long signature in
+// `from` using the RSA key `rsa`. The caller is responsible for ensuring that
+// `to` points to a memory section large enough to hold the message digest
+// (which is smaller than `RSA_size(rsa) - 11`). `padding` is the padding mode
+// that was used to sign the data. (See the OpenSSL man page for
+// RSA_private_decrypt(3) for more information.)
+//
+// Returns the length of `to` on success; otherwise, returns -1.
+//
+// The function signature comes from the prototype for `RSA_meth_set_priv_dec`
+// from the OpenSSL API.
+int PrivateDecrypt(int from_length, const unsigned char *from,
+                   unsigned char *to, RSA *rsa, int padding) {
+  // TODO(https://github.com/googleinterns/cloud-kms-oss-tools/issues/102):
+  // Currently unimplemented on purpose, but can be defined when keys with the
+  // `ASYMMETRIC_DECRYPT` purpose should be supported by the engine.
+  KMSENGINE_SIGNAL_ERROR(
+      Status(StatusCode::kUnimplemented, "Unsupported operation"));
+  return -1;
+}
+
+}  // namespace
+
+OpenSslRsaMethod MakeKmsRsaMethod() {
```

Comment:
I'm wondering whether this should return a `StatusOr` so that the failure condition is clearer.

Reply:
It's a little tricky here since looking at the source code of `RSA_meth_set_*`, none of these functions can actually ever fail. But nevertheless, it is probably a good idea to make this a `StatusOr` since otherwise the calling code needs to check for non-null.